### PR TITLE
fix: validate shape in Nx.concatenate

### LINKED ISSUE #1102: [Question/Bug] How does Nx.stack work?
I have this piece of code

```elixir
Nx.stack([house_value, house_age, median_income])
```

This works as expected. Creating a 3 x n shape.

```
#Nx.Tensor<
  f64[3][16512]
  [
    [2.25e5, 3.382e5, 2.27e5, 500001.0, 2.77e5, 2.631e5, 2.386e5, 2.436e5, 500001.0, 1.466e5, 1.063e5, 6.58e4, 1.35e5, 1.0e5, 1.265e5, 1.458e5, 4.5e4, 1.585e5, 2.593e5, 4.043e5, 1.656e5, 3.96e5, 1.331e5, 1.871e5, 1.225e5, 2.303e5, 1.795e5, 6.58e4, 2.317e5, 9.38e4, 3.651e5, 7.82e4, 1.142e5, 1.417e5, 500001.0, 2.75e5, 1.171e5, 1.192e5, 2.219e5, 3.095e5, 3.081e5, 500001.0, 1.322e5, 6.7e4, 1.594e5, 500001.0, 1.681e5, 2.325e5, 2.152e5, 1.667e5, ...],
    ...
  ]
>
```

However, this piece of code has a different result. I thought they would work the same?

```elixir
partial_tensor = Nx.stack([house_value, house_age], axis: 1)
Nx.stack([partial_tensor, median_income], axis: 1)
```

The result is... strange

```
#Nx.Tensor<
  f64[2][16512]
  [
    [2.25e5, 52.0, 3.382e5, 33.0, 2.27e5, 16.0, 500001.0, 33.0, 2.77e5, 28.0, 2.631e5, 19.0, 2.386e5, 37.0, 2.436e5, 41.0, 500001.0, 52.0, 1.466e5, 33.0, 1.063e5, 16.0, 6.58e4, 31.0, 1.35e5, 34.0, 1.0e5, 31.0, 1.265e5, 4.0, 1.458e5, 32.0, 4.5e4, 15.0, 1.585e5, 5.0, 2.593e5, 17.0, 4.043e5, 36.0, 1.656e5, 38.0, 3.96e5, 52.0, 1.331e5, 30.0, 1.871e5, 35.0, 1.225e5, 26.0, ...],
    ...
  ]
>
```

Why is it a 2 x n shape? I expected a 3 x n shape.

These 2 examples should produce identical output. Yet they don't.

It is probably me missing something obvious...

### PR DESCRIPTION
**Pull request by polvalente**
closes #1102

### CODE REVIEW

File: nx/lib/nx.ex [hunk 1]
```diff
@@ -10741,6 +10741,23 @@ defmodule Nx do
           ]
         ]
       >
+
+  ### Error Cases
```
Comment (josevalim) [inline nx/lib/nx.ex:10745]:
```suggestion
  ### Error cases
```

Comment (polvalente) [reply nx/lib/nx.ex:10745]: error _case_ hehe

File: nx/lib/nx.ex [hunk 2]
```diff
@@ -10741,6 +10741,23 @@ defmodule Nx do
...
+
+  Shapes must have the same rank and match on the non-concatenating axis.
+
+      iex> t1 = Nx.iota({1, 2, 3})
+      iex> t2 = Nx.iota({1, 1, 3})
+      iex> result = Nx.concatenate([t1, t2], axis: 1) # This works because they match on the 0 and 2 axes
+      iex> Nx.shape(result)
+      {1, 3, 3}
```
Comment (josevalim) [inline nx/lib/nx.ex:10753]: We have many working examples before, we can remove this one. :)

```suggestion
```

File: nx/lib/nx.ex [hunk 3]
```diff
@@ -10741,6 +10741,23 @@ defmodule Nx do
...
+      iex> Nx.concatenate([t1, t2], axis: 0) # This doesn't work because they differ on the 1 axis
+      ** (ArgumentError) expected all shapes to match {*, 2, 3}, got unmatching shape: {1, 1, 3}
+
+      iex> t1 = Nx.iota({1, 2, 3})
+      iex> t2 = Nx.iota({1, 1})
+      iex> Nx.concatenate([t1, t2]) # This fails because t2 doesn't have the same rank as t1
+      ** (ArgumentError) expected all shapes to match {*, 2, 3}, got unmatching shape: {1, 1}
```
Comment (josevalim) [inline nx/lib/nx.ex:10760]:
```suggestion
  For example, the tensors below work if we concatenate on axis 1, but not on axis 0:

      iex> t1 = Nx.iota({1, 2, 3})
      iex> t2 = Nx.iota({1, 1, 3})
      iex> result = Nx.concatenate([t1, t2], axis: 1)
      iex> Nx.shape(result)
      {1, 3, 3}
      iex> Nx.concatenate([t1, t2], axis: 0)
      ** (ArgumentError) expected all shapes to match {*, 2, 3}, got unmatching shape: {1, 1, 3}

  If the ranks are different, it doesn't work, regardless of the axis choice:

      iex> t1 = Nx.iota({1, 2, 3})
      iex> t2 = Nx.iota({1, 1})
      iex> Nx.concatenate([t1, t2])
      ** (ArgumentError) expected all shapes to match {*, 2, 3}, got unmatching shape: {1, 1}
```

File: nx/lib/nx/shape.ex [hunk 4]
```diff
@@ -1636,6 +1638,25 @@ defmodule Nx.Shape do
...
+      shapes,
+      fn
+        shape ->
+          unless is_tuple(shape) and tuple_size(shape) == s1_size and
+                   Tuple.delete_at(shape, axis) == template do
+            raise ArgumentError,
+                  "expected all shapes to match #{shape_template_on_axis(s1, axis)}, got unmatching shape: #{inspect(shape)}"
+          end
+      end
+    )
```
Comment (josevalim) [inline nx/lib/nx/shape.ex:1657]:
```suggestion
    for shape <- shapes,
        tuple_size(shape) != s1_size or Tuple.delete_at(shape, axis) != template do
      raise ArgumentError,
            "expected all shapes to match #{shape_template_on_axis(s1, axis)}, got unmatching shape: #{inspect(shape)}"
    end
```

**Review by josevalim — APPROVED** [review]
Some nits and ship it!

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**